File: flatten_ndc_schemas.py
import argparse
import json
import os
import xml.etree.ElementTree as ET
from collections import deque
import xml.dom.minidom

# Namespaces commonly used in XSD
XSD_NS = "http://www.w3.org/2001/XMLSchema"
ET.register_namespace("xs", XSD_NS)

def get_qname(tag):
    if '}' in tag:
        return tag.split('}', 1)[1]
    return tag

def get_namespace(tag):
    if '}' in tag:
        return tag.split('}', 1)[0][1:]
    return ""

def split_qname(text):
    if ':' in text:
        return text.split(':', 1)
    return None, text

class SchemaFlattener:
    def __init__(self, source_dir):
        self.source_dir = source_dir
        self.schemas = {} # path -> key info (root, target_ns, prefix_map, global_defs)
        self.global_definitions = {} # (ns, name) -> element

    def load_schema(self, filename):
        path = os.path.join(self.source_dir, filename)
        path = os.path.abspath(path)
        
        if path in self.schemas:
            return self.schemas[path]

        try:
            tree = ET.parse(path)
            root = tree.getroot()
        except Exception as e:
            print(f"Error loading {path}: {e}")
            return None

        # Extract targetNamespace
        target_ns = root.get('targetNamespace', '')
        
        # Extract namespace prefix mappings declared on this element
        # ElementTree hides xmlns attributes, but we can inspect the file or root
        # Actually, for resolving type="prefix:Name", we need the map.
        # We can reconstruct a simple map from the root's current scope hopefully, 
        # or parse the file manually for xmlns attributes if needed. 
        # ET handles namespaces by expanding them. Accessing original prefixes is hard.
        # However, attribute values like type="cns:ErrorType" are just strings.
        # We need to know what 'cns' maps to in THIS file.
        
        prefix_map = {}
        for event, (prefix, uri) in ET.iterparse(path, events=['start-ns']):
            prefix_map[prefix] = uri
        
        # Collect global definitions (complexType, simpleType, element, group, attributeGroup)
        defs = {}
        for child in root:
            tag_local = get_qname(child.tag)
            name = child.get('name')
            if name and tag_local in ('complexType', 'simpleType', 'element', 'group', 'attributeGroup'):
                defs[(target_ns, name)] = child

        schema_info = {
            'root': root,
            'target_ns': target_ns,
            'prefix_map': prefix_map,
            'definitions': defs,
            'imports': []
        }
        self.schemas[path] = schema_info
        self.global_definitions.update(defs)

        # Handle imports/includes recursively
        for child in root:
            tag_local = get_qname(child.tag)
            if tag_local in ('import', 'include'):
                loc = child.get('schemaLocation')
                if loc:
                    self.load_schema(loc)
                    schema_info['imports'].append(os.path.abspath(os.path.join(self.source_dir, loc)))

        return schema_info

    def resolve_reference(self, value, current_ns, prefix_map):
        """Resolves a string like 'cns:Type' to (namespace, local_name)"""
        if not value:
            return None
        
        prefix, local_name = split_qname(value)
        if prefix:
            ns = prefix_map.get(prefix)
            if ns:
                return (ns, local_name)
            # Check standard XS prefix just in case it wasn't captured or is implicit
            if prefix == 'xs': 
                return (XSD_NS, local_name)
            
            # Fallback: maybe the prefix IS the standard prefixes used in these files
            # For IATA 21.3, likely 'cns' is common types. 
            return (None, local_name) 
        else:
            # No prefix, usually means targetNamespace
            return (current_ns, local_name)

    def find_used_definitions(self, main_path):
        """Performs tree shaking starting from the main schema's global elements."""
        main_schema = self.schemas[main_path]
        used_keys = set()
        queue = deque()

        # Start with all global elements in the MAIN schema (the message itself)
        # e.g. IATA_OrderViewRS
        for (ns, name), el in main_schema['definitions'].items():
            if ns == main_schema['target_ns']:
                # Only root elements of THIS file are the entry points
                # (OR all globals in this file? typically the main file has one root element)
                # But let's assume everything defined in the main file is "used"
                queue.append((ns, name))
                used_keys.add((ns, name))

        while queue:
            current_key = queue.popleft()
            if current_key not in self.global_definitions:
                continue
            
            element = self.global_definitions[current_key]
            
            # Find the schema context where this element was defined 
            # (to resolve its prefixes correctly)
            # This is slow, so optimization: store origin in global_definitions or loop
            origin_schema = None
            for path, s in self.schemas.items():
                if current_key in s['definitions'] and s['definitions'][current_key] == element:
                    origin_schema = s
                    break
            
            if not origin_schema:
                continue

            # Traverse attributes and children for refs
            # We look for: type=, base=, ref=, itemType=, substitutionGroup=
            
            def check_ref(value):
                ref_key = self.resolve_reference(value, origin_schema['target_ns'], origin_schema['prefix_map'])
                # Only recursively add if it's not standard XSD and we define it
                if ref_key and ref_key[0] != XSD_NS:
                    # If we can't find the NS, we might skip it or warn.
                    # But for now, try to find it.
                    if ref_key in self.global_definitions:
                        if ref_key not in used_keys:
                            used_keys.add(ref_key)
                            queue.append(ref_key)

            # Check attributes of the element itself
            for attr in ['type', 'base', 'ref', 'itemType', 'substitutionGroup']:
                if element.get(attr):
                    check_ref(element.get(attr))

            # Recurse into children elements
            for child in element.iter():
                for attr in ['type', 'base', 'ref', 'itemType', 'substitutionGroup']:
                    if child.get(attr):
                        check_ref(child.get(attr))
        
        return used_keys

    def collect_namespaces(self, used_keys):
        """Collect all unique namespaces from used definitions."""
        namespaces = set()
        for (ns, name) in used_keys:
            if ns and ns != XSD_NS:
                namespaces.add(ns)
        return namespaces

    def build_prefix_map(self, namespaces, main_ns):
        """Build a consistent prefix map for all namespaces."""
        prefix_map = {'xs': XSD_NS}
        prefix_counter = 1
        
        for ns in sorted(namespaces):  # Sort for deterministic output
            if ns == main_ns:
                # Use 'tns' for target namespace or leave unprefixed
                prefix_map['tns'] = ns
            else:
                # Find an existing prefix from loaded schemas or generate one
                found_prefix = None
                for schema_info in self.schemas.values():
                    for prefix, uri in schema_info['prefix_map'].items():
                        if uri == ns and prefix and prefix not in prefix_map:
                            found_prefix = prefix
                            break
                    if found_prefix:
                        break
                
                if found_prefix:
                    prefix_map[found_prefix] = ns
                else:
                    prefix_map[f'ns{prefix_counter}'] = ns
                    prefix_counter += 1
        
        return prefix_map

    def get_prefix_for_ns(self, ns, output_prefix_map):
        """Get the prefix for a namespace in the output prefix map."""
        for prefix, uri in output_prefix_map.items():
            if uri == ns:
                return prefix
        return None

    def _write_schema_file(self, root, target_ns, output_path, xmlns_decls=None):
        """Write a schema element tree to a file with proper formatting."""
        xml_str = ET.tostring(root, encoding='utf-8').decode('utf-8')
        
        # Add any extra xmlns declarations
        if xmlns_decls:
            schema_tag_end = xml_str.find('>')
            if schema_tag_end > 0:
                xml_str = xml_str[:schema_tag_end] + ' ' + ' '.join(xmlns_decls) + xml_str[schema_tag_end:]
        
        # Pretty print with minidom
        dom = xml.dom.minidom.parseString(xml_str.encode('utf-8'))
        pretty_xml = dom.toprettyxml(indent="  ")
        
        # Remove empty lines
        pretty_xml = "\n".join([line for line in pretty_xml.split('\n') if line.strip()])
        
        with open(output_path, 'w') as f:
            f.write(pretty_xml)
        print(f"Generated {output_path}")

    def flatten_message(self, message_filename, output_path):
        """Flatten a message schema into two files: main + common types."""
        import copy
        
        main_info = self.load_schema(message_filename)
        if not main_info:
            return

        used_keys = self.find_used_definitions(os.path.abspath(os.path.join(self.source_dir, message_filename)))
        main_ns = main_info['target_ns']
        
        # Separate definitions by namespace
        defs_by_ns = {}
        for key in used_keys:
            ns, name = key
            if ns not in defs_by_ns:
                defs_by_ns[ns] = []
            defs_by_ns[ns].append(key)
        
        # Identify namespaces
        other_namespaces = [ns for ns in defs_by_ns.keys() if ns != main_ns and ns != XSD_NS]
        
        # Build prefix map for cross-references
        prefix_map = {'xs': XSD_NS}
        if main_ns:
            prefix_map[''] = main_ns  # default namespace for main schema types
        
        # Assign prefixes for other namespaces (typically just 'cns' for common types)
        for ns in other_namespaces:
            # Try to find existing prefix from source schemas
            found_prefix = None
            for schema_info in self.schemas.values():
                for pfx, uri in schema_info['prefix_map'].items():
                    if uri == ns and pfx:
                        found_prefix = pfx
                        break
                if found_prefix:
                    break
            prefix_map[found_prefix or 'cns'] = ns
        
        # Get prefix for common types namespace
        cns_prefix = None
        cns_ns = None
        for pfx, ns in prefix_map.items():
            if ns != main_ns and ns != XSD_NS and pfx:
                cns_prefix = pfx
                cns_ns = ns
                break

        # Helper to rewrite prefixes consistently
        # is_cns_schema: True when writing CommonTypes schema (needs cns: prefix for local types)
        def rewrite_prefixes(elem, origin_schema, target_schema_ns, is_cns_schema=False):
            for attr in ['type', 'base', 'ref', 'itemType', 'substitutionGroup']:
                val = elem.get(attr)
                if val:
                    ref_key = self.resolve_reference(val, origin_schema['target_ns'], origin_schema['prefix_map'])
                    if ref_key:
                        ref_ns, ref_local = ref_key
                        if ref_ns == XSD_NS:
                            elem.set(attr, f'xs:{ref_local}')
                        elif ref_ns == cns_ns and cns_prefix:
                            # Reference to common types - always use cns: prefix
                            elem.set(attr, f'{cns_prefix}:{ref_local}')
                        elif ref_ns == main_ns:
                            if is_cns_schema:
                                # From CommonTypes to main ns - need main ns prefix
                                elem.set(attr, f'msg:{ref_local}')
                            else:
                                # Same namespace (main) - no prefix needed
                                elem.set(attr, ref_local)
                        else:
                            elem.set(attr, ref_local)
            for child in elem:
                rewrite_prefixes(child, origin_schema, target_schema_ns, is_cns_schema)

        # Determine output filenames
        base_name = os.path.splitext(os.path.basename(output_path))[0]
        output_dir = os.path.dirname(output_path)
        common_types_filename = f"{base_name}_CommonTypes.xsd"
        common_types_path = os.path.join(output_dir, common_types_filename)

        # === Generate Common Types file (if there are types from other namespaces) ===
        if cns_ns and cns_ns in defs_by_ns:
            ET.register_namespace('xs', XSD_NS)
            ET.register_namespace(cns_prefix, cns_ns)  # Register cns for self-references
            
            cns_root = ET.Element(f'{{{XSD_NS}}}schema')
            cns_root.set('targetNamespace', cns_ns)
            cns_root.set('elementFormDefault', 'qualified')
            cns_root.set('version', '1.0')
            
            sorted_cns_keys = sorted(defs_by_ns[cns_ns], key=lambda x: x[1])
            
            for key in sorted_cns_keys:
                original_el = self.global_definitions[key]
                
                # Find origin schema
                origin_schema = None
                for path, s in self.schemas.items():
                    if key in s['definitions'] and s['definitions'][key] == original_el:
                        origin_schema = s
                        break
                if not origin_schema:
                    continue
                
                new_el = copy.deepcopy(original_el)
                rewrite_prefixes(new_el, origin_schema, cns_ns, is_cns_schema=True)
                cns_root.append(new_el)
            
            # Add xmlns:cns declaration for self-references
            xmlns_decls = [f'xmlns:{cns_prefix}="{cns_ns}"']
            self._write_schema_file(cns_root, cns_ns, common_types_path, xmlns_decls=xmlns_decls)

        # === Generate Main Message file ===
        ET.register_namespace('xs', XSD_NS)
        if cns_prefix and cns_ns:
            ET.register_namespace(cns_prefix, cns_ns)
        
        main_root = ET.Element(f'{{{XSD_NS}}}schema')
        main_root.set('targetNamespace', main_ns)
        main_root.set('elementFormDefault', 'qualified')
        main_root.set('version', main_info['root'].get('version', '1.0'))
        
        # Add import for common types
        if cns_ns and cns_ns in defs_by_ns:
            import_elem = ET.SubElement(main_root, f'{{{XSD_NS}}}import')
            import_elem.set('namespace', cns_ns)
            import_elem.set('schemaLocation', common_types_filename)
        
        # Add main namespace definitions
        if main_ns in defs_by_ns:
            sorted_main_keys = sorted(defs_by_ns[main_ns], key=lambda x: x[1])
            
            for key in sorted_main_keys:
                original_el = self.global_definitions[key]
                
                origin_schema = None
                for path, s in self.schemas.items():
                    if key in s['definitions'] and s['definitions'][key] == original_el:
                        origin_schema = s
                        break
                if not origin_schema:
                    continue
                
                new_el = copy.deepcopy(original_el)
                rewrite_prefixes(new_el, origin_schema, main_ns)
                main_root.append(new_el)
        
        # Build xmlns declarations for prefixes used in attribute values
        xmlns_decls = []
        if cns_prefix and cns_ns:
            xmlns_decls.append(f'xmlns:{cns_prefix}="{cns_ns}"')
        
        self._write_schema_file(main_root, main_ns, output_path, xmlns_decls)

def main():
    parser = argparse.ArgumentParser(description='Flatten NDC Schemas')
    parser.add_argument('--input-dir', required=True, help='Path to schema directory')
    parser.add_argument('--output-dir', required=True, help='Path to output directory')
    parser.add_argument('--message-list', help='JSON file containing message list or comma separated list')
    parser.add_argument('--version', help='Version key for JSON (e.g. 21.3.5)')

    args = parser.parse_args()

    if not os.path.exists(args.output_dir):
        os.makedirs(args.output_dir)

    target_files = []
    if args.message_list:
        if args.message_list.endswith('.json'):
            with open(args.message_list, 'r') as f:
                data = json.load(f)
                if args.version and 'versions' in data and args.version in data['versions']:
                    target_files = [m + ".xsd" for m in data['versions'][args.version]]
                else:
                    # Fallback or error
                    print("Version not found or invalid JSON structure")
                    return
        else:
            target_files = [m.strip() + ".xsd" if not m.endswith('.xsd') else m.strip() for m in args.message_list.split(',')]
    else:
        # Default: all top level XSDs in input dir?
        # Or just specific ones provided in prompt.
        # User prompt implies specific list "from iata_ndc_messages.json"
        pass

    if not target_files:
        print("No target files found.")
        return

    flattener = SchemaFlattener(args.input_dir)
    
    for fname in target_files:
        if not os.path.exists(os.path.join(args.input_dir, fname)):
            print(f"Skipping {fname} (not found)")
            continue
            
        out_name = os.path.join(args.output_dir, fname)
        flattener.flatten_message(fname, out_name)

if __name__ == '__main__':
    main()

File: tools/src/main/kotlin/com/ndc/tools/FlattenNdcSchemas.kt
package com.ndc.tools

import com.ndc.tools.common.NdcConstants
import kotlinx.serialization.json.*
import org.w3c.dom.Document
import org.w3c.dom.Element
import org.w3c.dom.Node
import java.io.File
import java.io.StringWriter
import java.util.ArrayDeque
import javax.xml.parsers.DocumentBuilderFactory
import javax.xml.transform.OutputKeys
import javax.xml.transform.TransformerFactory
import javax.xml.transform.dom.DOMSource
import javax.xml.transform.stream.StreamResult

private const val XSD_NS = "http://www.w3.org/2001/XMLSchema"

internal data class DefinitionKey(val ns: String, val name: String)

internal data class SchemaInfo(
    val doc: Document,
    val root: Element,
    val targetNs: String,
    val prefixMap: Map<String, String>, // prefix → URI
    val definitions: Map<DefinitionKey, Element>,
)

class SchemaFlattener(private val sourceDir: File) {
    private val schemas = mutableMapOf<File, SchemaInfo>() // absolute path → info
    private val globalDefinitions = mutableMapOf<DefinitionKey, Element>()
    private val externalSchemas = mutableMapOf<String, String>() // namespace → schemaLocation

    internal fun loadSchema(filename: String): SchemaInfo? {
        val path = sourceDir.resolve(filename).canonicalFile
        schemas[path]?.let { return it }

        val doc: Document
        try {
            val dbf = DocumentBuilderFactory.newInstance().apply {
                isNamespaceAware = true
                setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false)
            }
            doc = dbf.newDocumentBuilder().parse(path)
        } catch (e: Exception) {
            println("Error loading $path: ${e.message}")
            return null
        }

        val root = doc.documentElement
        val targetNs = root.getAttribute("targetNamespace") ?: ""

        // Extract prefix→URI map from the root element's attributes
        val prefixMap = mutableMapOf<String, String>()
        val attrs = root.attributes
        for (i in 0 until attrs.length) {
            val attr = attrs.item(i)
            if (attr.nodeName.startsWith("xmlns:")) {
                prefixMap[attr.nodeName.removePrefix("xmlns:")] = attr.nodeValue
            } else if (attr.nodeName == "xmlns") {
                prefixMap[""] = attr.nodeValue
            }
        }

        // Collect global definitions
        val defs = mutableMapOf<DefinitionKey, Element>()
        val defTags = setOf("complexType", "simpleType", "element", "group", "attributeGroup")
        var child = root.firstChild
        while (child != null) {
            if (child is Element && child.localName in defTags) {
                val name = child.getAttribute("name")
                if (name.isNotEmpty()) {
                    val key = DefinitionKey(targetNs, name)
                    defs[key] = child
                }
            }
            child = child.nextSibling
        }

        val info = SchemaInfo(doc, root, targetNs, prefixMap, defs)
        schemas[path] = info
        globalDefinitions.putAll(defs)

        // Handle imports/includes recursively
        child = root.firstChild
        while (child != null) {
            if (child is Element && child.localName in listOf("import", "include")) {
                val loc = child.getAttribute("schemaLocation")
                val ns = child.getAttribute("namespace")
                if (loc.isNotEmpty()) {
                    if (ns.isNotEmpty() && !ns.startsWith("http://www.iata.org")) {
                        externalSchemas.putIfAbsent(ns, loc)
                    } else {
                        loadSchema(loc)
                    }
                }
            }
            child = child.nextSibling
        }

        return info
    }

    private fun resolveReference(value: String, currentNs: String, prefixMap: Map<String, String>): DefinitionKey? {
        if (value.isEmpty()) return null
        val colonIdx = value.indexOf(':')
        if (colonIdx >= 0) {
            val prefix = value.substring(0, colonIdx)
            val localName = value.substring(colonIdx + 1)
            val ns = prefixMap[prefix]
            if (ns != null) return DefinitionKey(ns, localName)
            if (prefix == "xs") return DefinitionKey(XSD_NS, localName)
            return DefinitionKey("", localName)
        }
        return DefinitionKey(currentNs, value)
    }

    internal fun findUsedDefinitions(mainPath: File): Set<DefinitionKey> {
        val mainSchema = schemas[mainPath] ?: return emptySet()
        val usedKeys = mutableSetOf<DefinitionKey>()
        val queue = ArrayDeque<DefinitionKey>()

        // Seed with all global definitions in the main schema
        for (key in mainSchema.definitions.keys) {
            if (key.ns == mainSchema.targetNs) {
                queue.add(key)
                usedKeys.add(key)
            }
        }

        val refAttrs = listOf("type", "base", "ref", "itemType", "substitutionGroup")

        while (queue.isNotEmpty()) {
            val currentKey = queue.poll()
            val element = globalDefinitions[currentKey] ?: continue

            // Find origin schema for prefix resolution
            val originSchema = schemas.values.firstOrNull {
                it.definitions[currentKey] === element
            } ?: continue

            // Scan element and all descendants for references
            fun checkRef(value: String) {
                val refKey = resolveReference(value, originSchema.targetNs, originSchema.prefixMap) ?: return
                if (refKey.ns == XSD_NS) return
                if (refKey in globalDefinitions && refKey !in usedKeys) {
                    usedKeys.add(refKey)
                    queue.add(refKey)
                }
            }

            forEachElement(element) { el ->
                for (attr in refAttrs) {
                    val v = el.getAttribute(attr)
                    if (v.isNotEmpty()) checkRef(v)
                }
            }
        }

        return usedKeys
    }

    fun flattenMessage(messageFilename: String, outputPath: File) {
        val mainInfo = loadSchema(messageFilename) ?: return
        val mainPath = sourceDir.resolve(messageFilename).canonicalFile

        val usedKeys = findUsedDefinitions(mainPath)
        val mainNs = mainInfo.targetNs

        // Separate definitions by namespace
        val defsByNs = mutableMapOf<String, MutableList<DefinitionKey>>()
        for (key in usedKeys) {
            defsByNs.getOrPut(key.ns) { mutableListOf() }.add(key)
        }

        val otherNamespaces = defsByNs.keys.filter { it != mainNs && it != XSD_NS }

        // Build prefix map for common types namespaces
        val cnsNamespaces = mutableMapOf<String, String>() // namespace → prefix
        for (ns in otherNamespaces) {
            val foundPrefix = schemas.values.firstNotNullOfOrNull { schema ->
                schema.prefixMap.entries.firstOrNull { it.value == ns && it.key.isNotEmpty() }?.key
            }
            val prefix = foundPrefix ?: "cns${cnsNamespaces.size}"
            cnsNamespaces[ns] = prefix
        }

        // Scan for used external namespaces
        val usedExternalNs = mutableSetOf<String>()
        val refAttrs = listOf("type", "base", "ref", "itemType", "substitutionGroup")

        for (key in usedKeys) {
            val element = globalDefinitions[key] ?: continue
            val originSchema = schemas.values.firstOrNull { it.definitions[key] === element } ?: continue

            forEachElement(element) { el ->
                for (attr in refAttrs) {
                    val v = el.getAttribute(attr)
                    if (v.isNotEmpty()) {
                        val refKey = resolveReference(v, originSchema.targetNs, originSchema.prefixMap)
                        if (refKey != null && refKey.ns in externalSchemas) {
                            usedExternalNs.add(refKey.ns)
                        }
                    }
                }
            }
        }

        // Collect external namespace prefixes
        val externalNsPrefixes = mutableMapOf<String, String>()
        for (schema in schemas.values) {
            for ((pfx, uri) in schema.prefixMap) {
                if (uri in usedExternalNs && uri !in cnsNamespaces && uri != mainNs && uri != XSD_NS && pfx.isNotEmpty()) {
                    externalNsPrefixes.putIfAbsent(uri, pfx)
                }
            }
        }

        val baseName = outputPath.nameWithoutExtension
        val outputDir = outputPath.parentFile

        fun getCTSuffix(ns: String) = if ("FullyOptional" in ns) "_OptionalCommonTypes.xsd" else "_CommonTypes.xsd"

        // Helper: rewrite prefixes in a cloned element
        fun rewritePrefixes(elem: Element, originSchema: SchemaInfo, isCnsSchema: Boolean) {
            forEachElement(elem) { el ->
                for (attr in refAttrs) {
                    val v = el.getAttribute(attr)
                    if (v.isEmpty()) continue
                    val refKey = resolveReference(v, originSchema.targetNs, originSchema.prefixMap) ?: continue

                    val newValue = when {
                        refKey.ns == XSD_NS -> "xs:${refKey.name}"
                        refKey.ns in cnsNamespaces -> "${cnsNamespaces[refKey.ns]}:${refKey.name}"
                        refKey.ns == mainNs -> if (isCnsSchema) "msg:${refKey.name}" else refKey.name
                        refKey.ns in externalNsPrefixes -> "${externalNsPrefixes[refKey.ns]}:${refKey.name}"
                        else -> refKey.name
                    }
                    el.setAttribute(attr, newValue)
                }
            }
        }

        val outDoc = DocumentBuilderFactory.newInstance().apply { isNamespaceAware = true }
            .newDocumentBuilder().newDocument()

        // === Generate CommonTypes files ===
        for ((cnsTargetNs, cnsTargetPrefix) in cnsNamespaces) {
            if (cnsTargetNs !in defsByNs) continue

            val ctDoc = DocumentBuilderFactory.newInstance().apply { isNamespaceAware = true }
                .newDocumentBuilder().newDocument()
            val ctRoot = ctDoc.createElementNS(XSD_NS, "xs:schema")
            ctRoot.setAttribute("targetNamespace", cnsTargetNs)
            ctRoot.setAttribute("elementFormDefault", "qualified")
            ctRoot.setAttribute("version", "1.0")
            ctRoot.setAttribute("xmlns:xs", XSD_NS)
            ctRoot.setAttribute("xmlns:$cnsTargetPrefix", cnsTargetNs)
            ctDoc.appendChild(ctRoot)

            val usedExtNs = mutableSetOf<String>()
            val usedOtherCns = mutableSetOf<String>()

            val sortedKeys = defsByNs[cnsTargetNs]!!.sortedBy { it.name }
            for (key in sortedKeys) {
                val original = globalDefinitions[key] ?: continue
                val originSchema = schemas.values.firstOrNull { it.definitions[key] === original } ?: continue

                // Check for external and other cns references
                forEachElement(original) { el ->
                    for (attr in refAttrs) {
                        val v = el.getAttribute(attr)
                        if (v.isEmpty()) continue
                        val rk = resolveReference(v, originSchema.targetNs, originSchema.prefixMap) ?: continue
                        when {
                            rk.ns in externalNsPrefixes -> usedExtNs.add(rk.ns)
                            rk.ns in cnsNamespaces && rk.ns != cnsTargetNs -> usedOtherCns.add(rk.ns)
                        }
                    }
                }

                val cloned = ctDoc.importNode(original, true) as Element
                rewritePrefixes(cloned, originSchema, isCnsSchema = true)
                ctRoot.appendChild(cloned)
            }

            // Add imports for external namespaces
            for (extNs in usedExtNs.sorted()) {
                val importEl = ctDoc.createElementNS(XSD_NS, "xs:import")
                importEl.setAttribute("namespace", extNs)
                // Find schemaLocation from original imports
                for (schema in schemas.values) {
                    var imp = schema.root.firstChild
                    while (imp != null) {
                        if (imp is Element && imp.localName == "import" && imp.getAttribute("namespace") == extNs) {
                            val sl = imp.getAttribute("schemaLocation")
                            if (sl.isNotEmpty()) importEl.setAttribute("schemaLocation", sl)
                        }
                        imp = imp.nextSibling
                    }
                }
                ctRoot.insertBefore(importEl, ctRoot.firstChild)
                ctRoot.setAttribute("xmlns:${externalNsPrefixes[extNs]}", extNs)
            }

            // Add imports for other common types namespaces
            for (otherCnsNs in usedOtherCns.sorted()) {
                val importEl = ctDoc.createElementNS(XSD_NS, "xs:import")
                importEl.setAttribute("namespace", otherCnsNs)
                importEl.setAttribute("schemaLocation", "$baseName${getCTSuffix(otherCnsNs)}")
                ctRoot.insertBefore(importEl, ctRoot.firstChild)
                ctRoot.setAttribute("xmlns:${cnsNamespaces[otherCnsNs]}", otherCnsNs)
            }

            val ctOutputPath = outputDir.resolve("$baseName${getCTSuffix(cnsTargetNs)}")
            writeSchemaFile(ctDoc, ctOutputPath)
        }

        // === Generate Main Message file ===
        val mainDoc = DocumentBuilderFactory.newInstance().apply { isNamespaceAware = true }
            .newDocumentBuilder().newDocument()
        val mainRoot = mainDoc.createElementNS(XSD_NS, "xs:schema")
        mainRoot.setAttribute("targetNamespace", mainNs)
        mainRoot.setAttribute("elementFormDefault", "qualified")
        mainRoot.setAttribute("version", mainInfo.root.getAttribute("version").ifEmpty { "1.0" })
        mainRoot.setAttribute("xmlns:xs", XSD_NS)
        mainRoot.setAttribute("xmlns", mainNs)
        mainDoc.appendChild(mainRoot)

        // Add xmlns declarations for all common types prefixes
        for ((ns, prefix) in cnsNamespaces) {
            if (prefix.isNotEmpty()) {
                mainRoot.setAttribute("xmlns:$prefix", ns)
            }
        }

        // Add imports for common types
        for ((ns, _) in cnsNamespaces) {
            if (ns in defsByNs) {
                val importEl = mainDoc.createElementNS(XSD_NS, "xs:import")
                importEl.setAttribute("namespace", ns)
                importEl.setAttribute("schemaLocation", "$baseName${getCTSuffix(ns)}")
                mainRoot.appendChild(importEl)
            }
        }

        // Add main namespace definitions
        if (mainNs in defsByNs) {
            val sortedKeys = defsByNs[mainNs]!!.sortedBy { it.name }
            for (key in sortedKeys) {
                val original = globalDefinitions[key] ?: continue
                val originSchema = schemas.values.firstOrNull { it.definitions[key] === original } ?: continue

                val cloned = mainDoc.importNode(original, true) as Element
                rewritePrefixes(cloned, originSchema, isCnsSchema = false)
                mainRoot.appendChild(cloned)
            }
        }

        writeSchemaFile(mainDoc, outputPath)

        // Copy external schema files
        for (ns in usedExternalNs) {
            val loc = externalSchemas[ns] ?: continue
            val src = sourceDir.resolve(loc)
            val dst = outputDir.resolve(loc)
            if (src.exists() && !dst.exists()) {
                src.copyTo(dst)
                println("Copied external schema: $loc")
            }
        }
    }

    /** Clear state for processing a new message (keeps loaded schemas for reuse) */
    fun reset() {
        // Keep schemas loaded but clear global defs and externals
        // Actually, for batch processing of different versions, we need full reset
        schemas.clear()
        globalDefinitions.clear()
        externalSchemas.clear()
    }
}

private fun forEachElement(node: Node, action: (Element) -> Unit) {
    if (node is Element) action(node)
    var child = node.firstChild
    while (child != null) {
        forEachElement(child, action)
        child = child.nextSibling
    }
}

private fun writeSchemaFile(doc: Document, outputPath: File) {
    val transformer = TransformerFactory.newInstance().newTransformer().apply {
        setOutputProperty(OutputKeys.INDENT, "yes")
        setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2")
        setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "no")
        setOutputProperty(OutputKeys.ENCODING, "UTF-8")
    }

    val writer = StringWriter()
    transformer.transform(DOMSource(doc), StreamResult(writer))

    // Clean up: remove empty lines
    val result = writer.toString()
        .lines()
        .filter { it.isNotBlank() }
        .joinToString("\n")

    outputPath.writeText(result)
    println("Generated $outputPath")
}

// === Batch flatten (replaces batch_flatten.py) ===

private fun findMatchingRawFolder(version: String, rawFolders: List<String>): String? {
    return rawFolders.firstOrNull { folder ->
        folder == version || folder.startsWith("$version.") || folder.startsWith("${version}_")
    }
}

fun main(args: Array<String>) {
    // Support both single-version mode and batch mode
    var inputDir: String? = null
    var outputDir: String? = null
    var messageList: String? = null
    var version: String? = null

    for (arg in args) {
        when {
            arg.startsWith("--input-dir=") -> inputDir = arg.removePrefix("--input-dir=")
            arg.startsWith("--output-dir=") -> outputDir = arg.removePrefix("--output-dir=")
            arg.startsWith("--message-list=") -> messageList = arg.removePrefix("--message-list=")
            arg.startsWith("--version=") -> version = arg.removePrefix("--version=")
        }
    }

    if (inputDir != null && outputDir != null) {
        // Single-version mode (like flatten_ndc_schemas.py)
        flattenSingleVersion(File(inputDir), File(outputDir), messageList, version)
    } else {
        // Batch mode (like batch_flatten.py)
        batchFlatten()
    }
}

private fun flattenSingleVersion(inputDir: File, outputDir: File, messageListArg: String?, versionArg: String?) {
    outputDir.mkdirs()

    val targetFiles = mutableListOf<String>()
    if (messageListArg != null) {
        if (messageListArg.endsWith(".json")) {
            val data = Json.parseToJsonElement(File(messageListArg).readText()).jsonObject
            if (versionArg != null && "versions" in data) {
                val versions = data["versions"]!!.jsonObject
                if (versionArg in versions) {
                    targetFiles.addAll(versions[versionArg]!!.jsonArray.map { it.jsonPrimitive.content + ".xsd" })
                } else {
                    println("Version not found or invalid JSON structure")
                    return
                }
            }
        } else {
            targetFiles.addAll(messageListArg.split(",").map {
                val s = it.trim()
                if (s.endsWith(".xsd")) s else "$s.xsd"
            })
        }
    }

    if (targetFiles.isEmpty()) {
        println("No target files found.")
        return
    }

    val flattener = SchemaFlattener(inputDir)

    for (fname in targetFiles) {
        if (!inputDir.resolve(fname).exists()) {
            println("Skipping $fname (not found)")
            continue
        }

        val msgName = fname.removeSuffix(".xsd")
        val folderName = if (msgName.startsWith("IATA_")) msgName.removePrefix("IATA_") else msgName

        val msgOutputDir = outputDir.resolve(folderName)
        msgOutputDir.mkdirs()

        flattener.flattenMessage(fname, msgOutputDir.resolve(fname))
    }
}

private fun batchFlatten() {
    val projectRoot = NdcConstants.projectRoot()
    val jsonPath = projectRoot.resolve("iata_ndc_messages.json")
    val rawDir = projectRoot.resolve("raw_ndc_schemas")
    val outputBaseDir = projectRoot.resolve("ndc_schemas")

    if (!jsonPath.exists()) {
        println("Error: $jsonPath not found.")
        return
    }

    val data = Json.parseToJsonElement(jsonPath.readText()).jsonObject
    val versions = data["versions"]?.jsonObject?.keys ?: return

    val rawFolders = rawDir.listFiles()?.filter { it.isDirectory }?.map { it.name } ?: emptyList()

    for (version in versions) {
        val match = findMatchingRawFolder(version, rawFolders)
        if (match == null) {
            println("Skipping version $version: No matching directory found in $rawDir")
            continue
        }

        val inputDir = rawDir.resolve(match)
        val outputDir = outputBaseDir.resolve(version)

        println("--- Processing version $version (from $match) ---")

        flattenSingleVersion(inputDir, outputDir, jsonPath.absolutePath, version)
        println("Successfully flattened schemas for version $version into $outputDir")
    }
}
